### PR TITLE
feat: Add a custom encoder in Serializer

### PR DESCRIPTION
**Pull request by saranshdhingra**
While instantiating a Serializer we can pass a custom encoder as the last argument.
When present, a proto will be converted into an array using the custom function(encoder) supplied instead of the `encodeMessage` method in the `Serializer` class.

Essentially this provides a way to short-circuit the `encodeMessage` method.

### CODE REVIEW

File: tests/Tests/Unit/SerializerTest.php [hunk 1]
```diff
@@ -291,4 +291,50 @@ public function testSpecialEncodingDecodingByFieldType()
...
+            'red' => 100,
+            'blue' => 100
+        ];
+
+        return  [
+            [
+                [
+                    Color::class => function ($message) use($modified) {
+                        return $modified;
+                    }
```
Comment (bshaffer) [inline tests/Tests/Unit/SerializerTest.php:325]: nit, but can we make this serializer do something with the message and fields? This example is a little too contrived for me. Something like this:

```php
Color::class => function (Color $message) {
    return ['red' => $message->getRed() * 2, 'blue' => $message->getBlue() * 2];
}
```

File: tests/Tests/Unit/SerializerTest.php [hunk 2]
```diff
@@ -291,4 +291,50 @@ public function testSpecialEncodingDecodingByFieldType()
...
+                ],
+                new Color($original),
+                $modified
+            ],
+            [
+                // When no custom encoder is supplied, the encodeMessage will return the data
```
Comment (bshaffer) [inline tests/Tests/Unit/SerializerTest.php:331]: another nit, but in addition to supplying NO encoder, I'd like to see a test where an encoder is supplied for a _different_ message class, to illustrate that only matching classes are affected.

**Review by bshaffer — APPROVED** [review]
two minor requests, otherwise this looks great!

File: tests/Tests/Unit/SerializerTest.php (hunk 2 above)
Comment (saranshdhingra) [reply tests/Tests/Unit/SerializerTest.php:331]: Yeah, good idea.

**Pull request merged**